build_tools/python_deploy/build_linux_packages.sh

```diff
@@ -97,6 +165,122 @@ function run_in_docker() {
   done
 }
 
+
+function build_in_tree() {
+  local torch_from_src="$1"
+  local python_version="$2"
+  echo ":::: Build in-tree Torch from source: $torch_from_src with Python: $python_version"
+  cmake -GNinja -B/main_checkout/torch-mlir/build \
+      -DCMAKE_BUILD_TYPE=Release \
+      -DCMAKE_C_COMPILER=clang \
+      -DCMAKE_CXX_COMPILER=clang++ \
+      -DCMAKE_LINKER=lld \
+      -DLLVM_ENABLE_ASSERTIONS=ON \
+      -DCMAKE_C_COMPILER_LAUNCHER=ccache \
+      -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
+      -DLLVM_ENABLE_PROJECTS=mlir \
+      -DLLVM_EXTERNAL_PROJECTS="torch-mlir;torch-mlir-dialects" \
+      -DLLVM_EXTERNAL_TORCH_MLIR_SOURCE_DIR="/main_checkout/torch-mlir" \
+      -DLLVM_EXTERNAL_TORCH_MLIR_DIALECTS_SOURCE_DIR="/main_checkout/torch-mlir/externals/llvm-external-projects/torch-mlir-dialects" \
+      -DLLVM_TARGETS_TO_BUILD=host \
+      -DMLIR_ENABLE_BINDINGS_PYTHON=ON \
+      -DTORCH_MLIR_ENABLE_LTC=OFF \
+      -DTORCH_MLIR_USE_INSTALLED_PYTORCH="$torch_from_src" \
+      -DPython3_EXECUTABLE="$(which python3)" \
+      /main_checkout/torch-mlir/externals/llvm-project/llvm
+  cmake --build /main_checkout/torch-mlir/build
+  ccache -s
+}
+
+function test_in_tree() {
+  echo ":::: Test in-tree"
+  cmake --build /main_checkout/torch-mlir/build --target check-torch-mlir-all
+
+  cd /main_checkout/torch-mlir/
+  export PYTHONPATH="/main_checkout/torch-mlir/build/tools/torch-mlir/python_packages/torch_mlir"
+
+  echo ":::: Run refbackend e2e integration tests"
+  python -m e2e_testing.torchscript.main --config=refbackend -v
+
+  echo ":::: Run eager_mode e2e integration tests"
+  python -m e2e_testing.torchscript.main --config=eager_mode -v
+
+  echo ":::: Run TOSA e2e integration tests"
+  python -m e2e_testing.torchscript.main --config=tosa -v
+
+  echo ":::: Run Lazy Tensor Core e2e integration tests"
+  # Temporarily disabled in top of main (https://github.com/llvm/torch-mlir/pull/1292)
+  #python -m e2e_testing.torchscript.main --config=lazy_tensor_core -v
+}
+
+function setup_venv() {
+  local python_version="$1"
+  echo ":::: Setting up VENV with Python: $python_version"
+  python3 -m venv /main_checkout/torch-mlir/docker_venv
+  source /main_checkout/torch-mlir/docker_venv/bin/activate
+
+  echo ":::: pip installing dependencies"
+  python3 -m pip install -r /main_checkout/torch-mlir/externals/llvm-project/mlir/python/requirements.txt
+  python3 -m pip install -r /main_checkout/torch-mlir/requirements.txt
+
+}
```

Comment:
Am I understanding correct that these are runtime setup steps and hence the cost is incurred for every invocation of `build_linux_packages.sh`? My worry is paying this cost for every local re-run could be avoided if these were folded into the docker build step.

Reply:
Since it is pip cached you breeze through this in seconds (see measured runtimes).  
```
:::: pip installing dependencies                                                                                                      
Requirement already satisfied: numpy in /main_checkout/torch-mlir/docker_venv/lib/python3.10/site-packages (from -r /main_checkout/torch-mlir/externals/llvm-project/mlir/python/requirements.txt (line 1)) (1.23.2)
Requirement already satisfied: pybind11>=2.8.0 in /main_checkout/torch-mlir/docker_venv/lib/python3.10/site-packages (from -r /main_checkout/torch-mlir/externals/llvm-project/mlir/python/requirements.txt (line 2)) (2.10.0)
....
```

My entire OOT + In-tree build takes 3mins if I touch one file. 
```
TM_PACKAGES="out-of-tree in-tree" TM_PYTHON_VERSIONS="cp310-cp310"   0.20s user 0.18s system 0% cpu 3:16.42 total
```

This also takes care of cases when someone makes a change to requirements.txt.  We don't want end users to have to `docker build` it is an unnecessary burden.